lib/explorer/shared.ex

```diff
@@ -130,7 +133,11 @@ defmodule Explorer.Shared do
   defp type(item, _type) when is_integer(item), do: :integer
   defp type(item, _type) when is_float(item), do: :float
   defp type(item, _type) when is_boolean(item), do: :boolean
-  defp type(item, _type) when is_binary(item), do: :string
+  defp type(item, :binary) when is_binary(item), do: :binary
+
+  defp type(item, _type) when is_binary(item),
+    do: if(String.valid?(item), do: :string, else: :binary)
```

Comment:
This operation is expensive. I think we should always default to string and users explicitly override if necessary.